[2.9] Update DAGs guide (#3519)
astronomer/docs: learn/airflow-dag-parameters.md

---

> title: "DAG-level parameters"

Reviewer: ```suggestion
title: "DAG-level parameters in Airflow" 
```

sidebar_label: "DAG parameters"
id: airflow-dag-parameters

> description: "Learn about all DAG-level parameters in Airflow."

Reviewer: ```suggestion
description: "Learn about all important DAG-level parameters in Airflow."
```

Author: It is actually pretty much all of them haha, but I still think saying `important` makes sense, even if just to future proof.

---

In Airflow, you can configure when and how your DAG runs by setting parameters in the DAG object. DAG-level parameters affect how the entire DAG behaves, as opposed to task-level parameters which only affect a single task or [Airflow configs](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html) which affect the entire Airflow instance. 

This guide covers all user-relevant DAG-level parameters in Airflow.

> ## Basic

Reviewer: ```suggestion
## Basic DAG-level parameters
```

> There are four basic DAG-level parameters. It is a best practice to always set these parameters:

Reviewer: ```suggestion
There are four basic DAG-level parameters. It is best practice to always set these parameters in any DAG:
```

| `start_date` | The date and time after which the DAG starts being scheduled. Note that the first actual run of the DAG may be later than this date depending on how you define the schedule. See [DAG scheduling and timetables in Airflow](scheduling-in-airflow.md) for more information. This parameter may be required depending on your Airflow version and `schedule`. |
| `schedule`   | The schedule for the DAG. There are many different ways to define a schedule, see [Scheduling in Airflow](scheduling-in-airflow.md) for more information. Defaults to `timedelta(days=1)`. This parameter replaces the deprecated `schedule_interval` and `timetable` parameters. |
| `catchup`    | Whether the scheduler should backfill all missed DAG runs between the current date and the start date when the DAG is unpaused. This parameter defaults to `True`. It is a best practice to always set it to `False` unless you specifically want to backfill missed DAG runs, see [Catchup](rerunning-dags.md#catchup) for more information. |

> ## UI-related

Reviewer: ```suggestion
## UI parameters
```

> Some parameters relate to adding information to the DAG or change its appearance in the Airflow UI:

Reviewer: ```suggestion
Some parameters add documentation to a DAG or change its appearance in the Airflow UI:
```

> Some parameters relate to adding information to the DAG or change its appearance in the Airflow UI:

Reviewer: Link to our DAG docs guide?

| Parameter                           | Description |
|-------------------------------------|-------------|
| `description`                       | A short string that is displayed in the Airflow UI next to the DAG name. |
| `doc_md`        | A string that is rendered as [DAG documentation](custom-airflow-ui-docs-tutorial.md) in the Airflow UI. Tip: use `__doc__` to use the docstring of the Python file. It is a best practice to give all your DAGs a descriptive DAG documentation. |
| `owner_links`   | A dictionary with the key being the DAG owner and the value being a URL to link to when clicking on the owner in the Airflow UI. Commonly used as a mailto link to the owner's email address. Note that the `owner` parameter is set at the task level, usually by defining it in the `default_args` dictionary. |
| `tags`          | A list of tags shown in the Airflow UI to help with filtering DAGs. |
| `default_view`  | The default view of the DAG in the Airflow UI. Defaults to `grid`. |
| `orientation`   | The orientation of the DAG graph in the Airflow UI. Defaults to `LR` (left to right). |

> ## Jinja templating

Reviewer: ```suggestion
## Jinja templating parameters
```

There are parameters that relate to [Jinja templating](templating.md), such as:

| Parameter                           | Description |
|-------------------------------------|-------------|
| `template_searchpath`               | A list of folders where Jinja looks for templates. The path of the DAG file is included by default. |
| `template_undefined`             | The behavior of Jinja when a variable is undefined. Defaults to [StrictUndefined](https://jinja.palletsprojects.com/en/3.0.x/api/#jinja2.StrictUndefined). |
| `render_template_as_native_obj`     | Whether to render Jinja templates as native Python objects instead of strings. Defaults to `False`. |
| `user_defined_macros`            | A dictionary of macros that are available in the DAG's Jinja templates. Use `user_defined_filters` to add filters and `jinja_environment_kwargs` for additional Jinja configuration. See [Macros: using custom functions and variables in templates](templating.md#macros-using-custom-functions-and-variables-in-templates). |

## Scaling

> Three other helpful parameters relate to scaling in Airflow. For more information see [Scaling Airflow to optimize performance](airflow-scaling-workers.md):

Reviewer: ```suggestion
Some parameters can be used to scale your DAG's resource usage in Airflow. See [Scaling Airflow to optimize performance](airflow-scaling-workers.md) for more information.
```

| Parameter                           | Description |
|-------------------------------------|-------------|
| `max_active_tasks`                  | The number of task instances allowed to run concurrently for all DAG runs of this DAG. This parameter replaces the deprecated `concurrency`. |
| `max_active_runs`                   | The number of active DAG runs allowed to run concurrently for this DAG. |
| `max_consecutive_failed_dag_runs`   | (experimental) The maximum number of consecutive failed DAG runs, after which the scheduler will disable this DAG. |

> ## Callbacks

Reviewer: ```suggestion
## Callback parameters
```

> 

Reviewer: ```suggestion

These parameters help you configure the behavior of [Airflow callbacks](error-notifications-in-airflow.md#airflow-callbacks).

```

| Parameter                           | Description |
|-------------------------------------|-------------|
| `on_success_callback`                  | A function to be executed after completion of a successful DAG run. |
| `on_failure_callback`                   | A function to be executed after a failed DAG run. |
| `sla_miss_callback`   | A function to be executed when a DAG misses its defined [Service Level Agreement (SLA)](error-notifications-in-airflow.md#airflow-service-level-agreements) |

> See [Airflow callbacks](error-notifications-in-airflow.md#airflow-callbacks) for more information.

Reviewer: ```suggestion
```

:::tip

> Astronomer customers can use Astro alerts instead of or in addition to Airflow callbacks. See [When to use Airflow or Astro alerts for your pipelines on Astro](https://docs.astronomer.io/astro/best-practices/airflow-vs-astro-alerts) for more information.

Reviewer: ```suggestion
On Astro, you can use Astro alerts instead of or in addition to Airflow callbacks. See [When to use Airflow or Astro alerts for your pipelines on Astro](https://docs.astronomer.io/astro/best-practices/airflow-vs-astro-alerts) for more information.
```

:::

> ## Other

Reviewer: ```suggestion
## Other parameters
```

| `render_template_as_native_obj`     | Whether to render Jinja templates as native Python objects instead of strings. Defaults to `False`. |
| `user_defined_macros`            | A dictionary of macros that are available in the DAG's Jinja templates. Use `user_defined_filters` to add filters and `jinja_environment_kwargs` for additional Jinja configuration. See [Macros: using custom functions and variables in templates](templating.md#macros-using-custom-functions-and-variables-in-templates). |

## Scaling

Some parameters can be used to scale your DAG's resource usage in Airflow. See [Scaling Airflow to optimize performance](airflow-scaling-workers.md) for more information.

| Parameter                           | Description |
|-------------------------------------|-------------|
| `max_active_tasks`                  | The number of task instances allowed to run concurrently for all DAG runs of this DAG. This parameter replaces the deprecated `concurrency`. |
| `max_active_runs`                   | The number of active DAG runs allowed to run concurrently for this DAG. |

> | `max_consecutive_failed_dag_runs`   | (experimental) The maximum number of consecutive failed DAG runs, after which the scheduler will disable this DAG. |

Author: ```suggestion
| `max_consecutive_failed_dag_runs`   | (experimental, added in Airflow 2.9) The maximum number of consecutive failed DAG runs, after which the scheduler will disable this DAG. |
```
